sheepdog/transactions/upload/sub_entities.py

```diff
@@ -0,0 +1,452 @@
+"""
+Subclasses for UploadEntity that handle different types of
+uploaded entites.
+"""
+import uuid
+import flask
+
+from sheepdog.transactions.entity_base import EntityErrors
+
+from sheepdog.transactions.upload.entity import UploadEntity
+from sheepdog.transactions.upload.entity import lookup_node
+
+
+class NonFileUploadEntity(UploadEntity):
+    """
+    Deals with non-file-like entities.
+
+    At the moment, there is no non-file specific logic for upload entities.
+
+    This was created for clarity and future extensibility in the case that
+    non-file-like upload entity logic is required.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(NonFileUploadEntity, self).__init__(*args, **kwargs)
+
+
+class FileUploadEntity(UploadEntity):
+    """
+    Deals with file-like entities.
+
+    As such, this class extends the general UploadEntity class by
+    handling the interaction to the index service.
+
+    Like it's parent class:
+    The lifespan of a FileUploadEntity should include:
+       #. parse
+       #. pre_validate
+       #. instantiate
+       #. post_validate
+    """
+    def __init__(self, *args, **kwargs):
+        super(FileUploadEntity, self).__init__(*args, **kwargs)
+        self.file_exists = False
+        self.file_index = None
+        self.file_by_uuid = None
+        self.file_by_hash = None
+        self.urls = []
+
+    def parse(self, doc):
+        """
+        Parse the given doc and set the instance values accordingly.
+
+        Args:
+            doc (dict): the json upload representation
+
+        Return:
+            None
+        """
+        # If a url is included, get it here and remove it from doc so it
+        # doesn't get validated
+        entity_urls = doc.get('urls')
+        if entity_urls:
+            self.urls = entity_urls.strip().split(',')
+            # remove from the doc since we don't want to validate it
+            doc.pop('urls')
+
+        super(FileUploadEntity, self).parse(doc)
+
+    def get_node_create(self, skip_node_lookup=False):
+        """
+        This is called for a POST operation.
+
+        Return:
+            psqlgraph.Node
+        """
+        # ################################################################
+        # SignpostClient is used instead of IndexClient for the GDCAPI.
+        # This means that the client doesn't have access to IndexClient's
+        # methods, causing exceptions to occur.
+        #
+        # Temporary workaround until gdcapi uses indexd
+        # ################################################################
+        if self._config.get('USE_SIGNPOST', False):
```

Comment:
I left previous behavior untouched to the best of my ability and added a few tests to test_upload_signpost.py where I make sure I didn't break previous behavior for this flag. Probably he should work off the changes in this branch since there was quite a bit of refactoring